_posts/2020-01-30-Age-Targeting-Part1.md

```diff
@@ -0,0 +1,129 @@
+Apps where the frac12 value is higher than 30% are excluded.
+Below you can see the distribution of the frac12 variable and the consequences of the rejection of the non reliable apps on the total age distribution.
+![frac12_cut](../images/Age-Targeting/frac12_cut.png)
+The red vertical line indicates the threshold we apply over the frac12 feature.
+
+
+* **Uniform distribution**<br/>
+One of the four examples (the bottom left one) clearly show a uniform generated distribution.
+Detect those apps is relatively easy, since for a fixed device a different age will be provided every time the device accesses the app.
+For every devices which have been observed at least twice in the app, we calcualte the max - min age value and, then, we calculate the average of this differences.
+In the image you can see the distribution of the avg_range_age variable and the effect of the cut on the global distributions.
+![frac12_cut](../images/Age-Targeting/uniform_cut.png)
+
+
+* **Symmetric like shapes**<br/>
+Certain distributions show a shape which seems to be suspicious symmetric.
+Using the [skewness](https://en.wikipedia.org/wiki/Skewness) statistical measure we can select those and drop them out.
+In the plot below you can see the distribution of the skewness and the effect of the cut on the global distributions.
+![frac12_cut](../images/Age-Targeting/skewness_cut.png)
+
+
+* **Too many peaks**<br/>
+Using a basic peak detecting algorithm, written in pyspark and reported below, we can count the number of peaks on the age distribution.
+```python
+from pyspark.sql import functions as F
+from pyspark.sql.types import IntegerType
+from pyspark.sql import Window
+def get_apps_with_peak_information(df):
+    window_frac = Window.partitionBy(df['apps']).orderBy(df['age'].asc())
+    df = df.withColumn("lag_1", F.lag("num_dev_per_age", 1).over(window_frac))
+    df = df.withColumn("lag_2", F.lag("num_dev_per_age", 2).over(window_frac))
+    df = df.withColumn("lag_3", F.lag("num_dev_per_age", 3).over(window_frac))
+    df = df.withColumn("lead_1", F.lead("num_dev_per_age", 1).over(window_frac))
+    df = df.withColumn("lead_2", F.lead("num_dev_per_age", 2).over(window_frac))
+    df = df.withColumn("lead_3", F.lead("num_dev_per_age", 3).over(window_frac))
+    df = df.withColumn(
+        "is_higher_than_nei",
+        ((df.num_dev_per_age > df.lag_1) &
+         (df.lag_1 > df.lag_2) &
+         (df.lag_2 > df.lag_3) &
+
+         (df.num_dev_per_age > df.lead_1) &
+         (df.lead_1 > df.lead_2) &
+         (df.lead_2 > df.lead_3)
+         ).cast(IntegerType())
+    )
+    df = df.withColumn("frac_for_peak", 
+                       (df.frac_num_age > peak_threshold).cast(IntegerType()))
+    df = df.withColumn("is_a_peak", df.frac_for_peak * df.is_higher_than_nei)
+    return df.groupby("apps").agg(F.sum("is_a_peak").alias("num_peaks"))
+```
```

Comment:
Why do you only show code for this one pattern? Why not for all of them?

Reply:
Because is the only one which I like and that I think is nice to show, should I add one for everything?

Reply:
I added code for every cut now